pkg/runtime/wfengine/workflowstate.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package wfengine
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/microsoft/durabletask-go/backend"
+
+	"github.com/dapr/dapr/pkg/actors"
+)
+
+const (
+	inboxKeyPrefix   = "inbox"
+	historyKeyPrefix = "history"
+	customStatusKey  = "customStatus"
+	metadataKey      = "metadata"
+)
+
+type workflowState struct {
+	Inbox        []*backend.HistoryEvent
+	History      []*backend.HistoryEvent
+	CustomStatus string
+	Generation   uuid.UUID
+
+	// change tracking
+	inboxAddedCount     int
+	inboxRemovedCount   int
+	historyAddedCount   int
+	historyRemovedCount int
+}
+
+type workflowStateMetadata struct {
+	InboxLength   int
+	HistoryLength int
+	Generation    uuid.UUID
+}
+
+func NewWorkflowState(generation uuid.UUID) *workflowState {
+	return &workflowState{
+		Generation: generation,
+	}
+}
+
+func (s *workflowState) Reset() {
+	s.inboxRemovedCount += len(s.Inbox)
+	s.Inbox = nil
+	s.historyRemovedCount += len(s.History)
+	s.History = nil
+	s.CustomStatus = ""
+	s.Generation = uuid.New()
+}
+
+func (s *workflowState) ApplyRuntimeStateChanges(runtimeState *backend.OrchestrationRuntimeState) {
+	if runtimeState.ContinuedAsNew() {
+		s.historyRemovedCount += len(s.History)
+		s.History = nil
```

Comment:
Need to cover this case.

Reply:
Done. I also fixed a bug that was exposed by the newly added test, so thanks for calling this out.